# Problema NEAREST NEIGHBORS

In [33]:
import pandas as pd
import numpy as np
from sklearn.neighbors import NearestNeighbors
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt

In [34]:
vino = pd.read_csv('wine-clustering.csv')
vino.head()

Alcohol  Malic_Acid   Ash  Ash_Alcanity  Magnesium  Total_Phenols  \
0    14.23        1.71  2.43          15.6        127           2.80   
1    13.20        1.78  2.14          11.2        100           2.65   
2    13.16        2.36  2.67          18.6        101           2.80   
3    14.37        1.95  2.50          16.8        113           3.85   
4    13.24        2.59  2.87          21.0        118           2.80   

   Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  Color_Intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   OD280  Proline  
0   3.92     1065  
1   3.40     1050  
2   3.17     1185  
3   3.45     1480  
4   2.93      735

In [45]:
test = pd.DataFrame({
    'Alcohol': [14],
    'Malic_Acid': [2],
    'Ash': [2.5],
    'Ash_Alcanity': [16],
    'Magnesium': [115],
    'Total_Phenols': [3],
    'Flavanoids': [2.5],
    'Nonflavanoid_Phenols': [0.4],
    'Proanthocyanins': [2],
    'Color_Intensity': [9],
    'Hue': [1],
    'OD280': [3.5],
    'Proline': [800]
})

test

Alcohol  Malic_Acid  Ash  Ash_Alcanity  Magnesium  Total_Phenols  \
0       14           2  2.5            16        115              3   

   Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  Color_Intensity  Hue  \
0         2.5                   0.4                2                9    1   

   OD280  Proline  
0    3.5      800

In [42]:
KNN_modelo_regresion = NearestNeighbors(n_neighbors = 5, algorithm='auto')
KNN_modelo_regresion.fit(vino)

NearestNeighbors()

In [59]:
# Encontrar los vecinos más cercanos
distancias, indices = KNN_modelo_regresion.kneighbors(test)

# Imprimir los resultados
print("Distancias a los vecinos más cercanos:", distancias)
print("Índices de los vecinos más cercanos:", indices, "\n")

# Mostrar los vinos más cercanos
vinos_mas_cercanos = pd.DataFrame(vino.iloc[indices[0]])
print("Los 5 vinos más cercanos:")
vinos_mas_cercanos

Distancias a los vecinos más cercanos: [[ 6.1946267  23.08566871 33.06187079 33.17827301 33.37464756]]
Índices de los vecinos más cercanos: [[ 40  20  25  21 141]] 

Los 5 vinos más cercanos:


Alcohol  Malic_Acid   Ash  Ash_Alcanity  Magnesium  Total_Phenols  \
40     13.56        1.71  2.31          16.2        117           3.15   
20     14.06        1.63  2.28          16.0        126           3.00   
25     13.05        2.05  3.22          25.0        124           2.63   
21     12.93        3.80  2.65          18.6        102           2.41   
141    13.36        2.56  2.35          20.0         89           1.40   

     Flavanoids  Nonflavanoid_Phenols  Proanthocyanins  Color_Intensity   Hue  \
40         3.29                  0.34             2.34             6.13  0.95   
20         3.17                  0.24             2.10             5.65  1.09   
25         2.68                  0.47             1.92             3.58  1.13   
21         2.41                  0.25             1.98             4.50  1.03   
141        0.50                  0.37             0.64             5.60  0.70   

     OD280  Proline  
40    3.38      795  
20    3.71      780  
25    3.20      830  
21    3.52      770  
141   2.47      780

# Problema MARKET BASQUET

In [192]:
mi_canasta_clientes = [['bread', 'butter', 'wine', 'bananas', 'coffee', 'carrots'],
    ['tomatoes', 'onions', 'cheese', 'milk', 'potatoes'],
    ['beer', 'chips', 'asparagus', 'salsa', 'milk', 'apples'],
    ['olive oil', 'bread', 'butter', 'tomatoes', 'steak', 'carrots'],
    ['tomatoes', 'onions', 'chips', 'wine', 'ketchup', 'orange juice'],
    ['bread', 'butter', 'beer', 'chips', 'milk'],
    ['butter', 'tomatoes', 'carrots', 'coffee', 'sugar'],
    ['tomatoes', 'onions', 'cheese', 'milk', 'potatoes'],
    ['bread', 'butter', 'ketchup', 'coffee', 'chicken wings'],
    ['butter', 'beer', 'chips', 'asparagus', 'apples'],
    ['tomatoes', 'onion', 'beer', 'chips', 'milk', 'coffee']]

In [200]:
import pandas as pd
import numpy as np

# Crear el DataFrame inicial
todos_los_productos = list(set(product for basket in mi_canasta_clientes for product in basket))
clientes = range(len(mi_canasta_clientes))

Tabla_clientes = pd.DataFrame(
    data=np.zeros((len(clientes), len(todos_los_productos)), dtype=int),
    columns=todos_los_productos,
    index=clientes
)

# Rellenar la tabla con un bucle
for cliente, productos in enumerate(mi_canasta_clientes):
    for producto in productos:
        Tabla_clientes.at[cliente, producto] = 1

# Mostrar la tabla resultante
Tabla_clientes


ketchup  orange juice  beer  asparagus  chicken wings  milk  butter  \
0         0             0     0          0              0     0       1   
1         0             0     0          0              0     1       0   
2         0             0     1          1              0     1       0   
3         0             0     0          0              0     0       1   
4         1             1     0          0              0     0       0   
5         0             0     1          0              0     1       1   
6         0             0     0          0              0     0       1   
7         0             0     0          0              0     1       0   
8         1             0     0          0              1     0       1   
9         0             0     1          1              0     0       1   
10        0             0     1          0              0     1       0   

    tomatoes  potatoes  bread  ...  cheese  olive oil  bananas  chips  \
0          0         0      1  ...       0          0        1      0   
1          1         1      0  ...       1          0        0      0   
2          0         0      0  ...       0          0        0      1   
3          1         0      1  ...       0          1        0      0   
4          1         0      0  ...       0          0        0      1   
5          0         0      1  ...       0          0        0      1   
6          1         0      0  ...       0          0        0      0   
7          1         1      0  ...       1          0        0      0   
8          0         0      1  ...       0          0        0      0   
9          0         0      0  ...       0          0        0      1   
10         1         0      0  ...       0          0        0      1   

    carrots  onion  onions  apples  wine  sugar  
0         1      0       0       0     1      0  
1         0      0       1       0     0      0  
2         0      0       0       1     0      0  
3         1      0       0       0     0      0  
4         0      0       1       0     1      0  
5         0      0       0       0     0      0  
6         1      0       0       0     0      1  
7         0      0       1       0     0      0  
8         0      0       0       0     0      0  
9         0      0       0       1     0      0  
10        0      1       0       0     0      0  

[11 rows x 23 columns]

### Revision extra para revisar integridad de los datos

In [216]:
# Bucle para cada cliente
for cliente in Tabla_clientes.index:  # Recorre cada índice del DataFrame (clientes)
    productos_comprados = Tabla_clientes.loc[cliente][Tabla_clientes.loc[cliente] == 1].index.tolist()
    print(f"Cliente {cliente} compró: {productos_comprados}")
    print(f"Los productos están íntegros: {set(productos_comprados) == set(mi_canasta_clientes[cliente])} ")

Cliente 0 compró: ['butter', 'bread', 'coffee', 'bananas', 'carrots', 'wine']
Los productos están íntegros: True
Cliente 1 compró: ['milk', 'tomatoes', 'potatoes', 'cheese', 'onions']
Los productos están íntegros: True
Cliente 2 compró: ['beer', 'asparagus', 'milk', 'salsa', 'chips', 'apples']
Los productos están íntegros: True
Cliente 3 compró: ['butter', 'tomatoes', 'bread', 'steak', 'olive oil', 'carrots']
Los productos están íntegros: True
Cliente 4 compró: ['ketchup', 'orange juice', 'tomatoes', 'chips', 'onions', 'wine']
Los productos están íntegros: True
Cliente 5 compró: ['beer', 'milk', 'butter', 'bread', 'chips']
Los productos están íntegros: True
Cliente 6 compró: ['butter', 'tomatoes', 'coffee', 'carrots', 'sugar']
Los productos están íntegros: True
Cliente 7 compró: ['milk', 'tomatoes', 'potatoes', 'cheese', 'onions']
Los productos están íntegros: True
Cliente 8 compró: ['ketchup', 'chicken wings', 'butter', 'bread', 'coffee']
Los productos están íntegros: True
Cliente 9 c

In [218]:
from mlxtend.frequent_patterns import apriori

In [220]:
compras_frecuentes = apriori(Tabla_clientes, min_support= 0.3, 
                             use_colnames = True ).sort_values('support', ascending = False).reset_index(drop=True)
compras_frecuentes['length']= compras_frecuentes['itemsets'].apply(lambda x: len(x))
compras_frecuentes

C:\Users\BrunoLap\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support         itemsets  length
0  0.545455         (butter)       1
1  0.545455       (tomatoes)       1
2  0.454545           (milk)       1
3  0.454545          (chips)       1
4  0.363636           (beer)       1
5  0.363636          (bread)       1
6  0.363636         (coffee)       1
7  0.363636    (chips, beer)       2
8  0.363636  (butter, bread)       2

## Explicacion:

### En los datos aparece mantequilla y tomates más del 50% quiere decir que lo compran recurrentemente, y que el dos veces 36% aparece el conjunto (cerveza, papas) y (mantequilla, pan)

In [221]:
from mlxtend.frequent_patterns import association_rules

In [228]:
association_rules(compras_frecuentes, metric= 'confidence', 
                             min_threshold = 0.5 ).sort_values('confidence', ascending = False).reset_index(drop=True)

antecedents consequents  antecedent support  consequent support   support  \
0      (beer)     (chips)            0.363636            0.454545  0.363636   
1     (bread)    (butter)            0.363636            0.545455  0.363636   
2     (chips)      (beer)            0.454545            0.363636  0.363636   
3    (butter)     (bread)            0.545455            0.363636  0.363636   

   confidence      lift  representativity  leverage  conviction  \
0    1.000000  2.200000               1.0  0.198347         inf   
1    1.000000  1.833333               1.0  0.165289         inf   
2    0.800000  2.200000               1.0  0.198347    3.181818   
3    0.666667  1.833333               1.0  0.165289    1.909091   

   zhangs_metric   jaccard  certainty  kulczynski  
0       0.857143  0.800000   1.000000    0.900000  
1       0.714286  0.666667   1.000000    0.833333  
2       1.000000  0.800000   0.685714    0.900000  
3       1.000000  0.666667   0.476190    0.833333

## Explicacion:

### El 100% de los casos donde compran cerveza tambien papas, e igual el 100% de las veces que se compra pan, se lleva mantequilla. El 80% si es viceversa, si se compran papas se lleva cerveza y si se lleva mantequilla se llevan pan pero con 66% de probabilidad.

### Es 2.2 veces más probable que compren papas si cerveza es comprada, mientras que en el pan es 1.8 veces más probable que se lleven mantequilla si llevan pan.